src/channeldef.ts

```diff
@@ -198,6 +199,72 @@ export type TextDef<F extends Field> =
   | FieldOrDatumDefWithCondition<StringDatumDef<F>, Text>
   | ValueDefWithCondition<StringFieldDef<F>, Text>;
 
+/**
+ * Position of a label relative to its base mark's position, which includes anchor direction (anchor) and distance (offset).
+ */
+export type LabelPosition = {
+  /**
+   * distance of a label from its base mark.
+   * if negative, the label position is insize.
+   */
+  offset: number;
+
+  /**
+   * anchor direction of a label from its base mark.
+   * The value could be one of `'left '` , `'right '` , `'top '` , `'middle '` ,
+   * `'bottom '` , `'top-left '` , `'top-right '` , `'bottom-left '` , or  '`bottom-right '`
+   */
+  anchor: LabelAnchor;
+};
+
+export type LabelDefMixins = {
+  /**
+   * A list of possible positions for a label to be placed relative to its base mark.
+   */
+  position?: LabelPosition[];
+
+  /**
+   * The layers of marks that the labels will avoid.
+   * `'all'` means all the marks that are children of the root layer.
+   * number `n` means all the marks that are children of the nth ancestor layer.
+   *
+   * __Default value:__ `0` (only avoid the base mark)
+   */
+  avoidParentLayer?: 'all' | number;
```

Comment:
I think this API sounds confusing.  

As it is described, this sounds more like avoidAncestorLayer (not parent, since parent should mean `1` only).

Reply:
What about just `avoid: 'all-marks' | 'base-mark'`?

(I wonder if we really need the children of the nth ancestor because that sounds confusing.)

Reply:
I think that `avoid: 'all-marks' | 'base-mark'` is better too. But, there is a problem when users use a `line` mark with `point: true`. Using `'base-mark'` before normalizing should mean both the line and the point. However, We do not have a way to represent that after we normalize the spec.
So, I decide to use `avoidParentLayer` (or `avoidAncestorLayer`) because we can increment the avoid level by one whenever we have a normalization for line with point.

Reply:
I'm not sure if I totally understand the following point yet:

> So, I decide to use avoidParentLayer (or avoidAncestorLayer) because we can increment the avoid level by one whenever we have a normalization for line with point.



But this sounds the the syntax is designed to optimize for a corner case.

If anything,  `avoid: 'all-marks' | 'base-mark' | {ancestor: number}`` would probably make it more readable.

Reply:
I see. The syntax you suggested `avoid: 'all-marks' | 'base-mark' | {ancestor: number}` sounds great to me. I will make the change.